File: algorithms/39.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
private:
    void dfs(vector< vector<int> > &result, vector<int> &combination, vector<int> &candidates, int index, int target)
    {
        if ( target == 0 )
        {
            result.push_back(combination);
            return;
        }
        for ( unsigned int i = index; i < candidates.size() && target >= candidates[i]; i++ )
        {
            combination.push_back(candidates[i]);
            dfs(result, combination, candidates, i, target-candidates[i]);
            combination.pop_back();
        }
        return;
    }
public:
    vector< vector<int> > combinationSum(vector<int>& candidates, int target)
    {
        vector< vector<int> > result;
        vector<int> combination;
        sort(candidates.begin(), candidates.end());
        dfs(result, combination, candidates, 0, target);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python/39.py
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        dp = [set() for i in range(target+1)]
        dp[0].add(())
        for item in candidates:
            for i in range(target+1):
                j = 1
                while i + j * item <= target:
                    k = i + j * item
                    dp[k].update(set(map(lambda x: x + (item, ) * j, dp[i])))
                    j += 1
        return list(dp[target])
